cpp/serve/request.h

```diff
@@ -0,0 +1,222 @@
+/*!
+ *  Copyright (c) 2023 by Contributors
+ * \file request.h
+ * \brief Implementation of llm chat.
+ */
+#ifndef MLC_LLM_SERVE_REQUEST_H_
+#define MLC_LLM_SERVE_REQUEST_H_
+
+#include <tvm/node/reflection.h>
+#include <tvm/node/structural_equal.h>
+#include <tvm/node/structural_hash.h>
+#include <tvm/runtime/container/array.h>
+#include <tvm/runtime/container/optional.h>
+#include <tvm/runtime/container/shape_tuple.h>
+#include <tvm/runtime/container/string.h>
+#include <tvm/runtime/object.h>
+
+#include "config.h"
+#include "input.h"
+
+namespace mlc {
+namespace llm {
+namespace serve {
+
+using namespace tvm::runtime;
+
+/****************** Request ******************/
+
+/*!
+ * \brief The user submitted text-generation request, which contains
+ * a list of multi-modal inputs and a set of sampling configuration
+ * parameters.
+ */
+class RequestNode : public Object {
+ public:
+  Array<Input> inputs;
+  /*! \brief JSON string containing temperature, top_p, repetition_penalty, max_gen_len, etc. */
+  SampleConfig sampling_params;
+
+  void VisitAttrs(tvm::AttrVisitor* v) {
+    v->Visit("inputs", &inputs);
+    v->Visit("sampling_params", &sampling_params);
+  }
+
+  bool SEqualReduce(const RequestNode* other, tvm::SEqualReducer equal) const {
+    return equal(inputs, other->inputs) && equal(sampling_params, other->sampling_params);
+  }
+
+  void SHashReduce(tvm::SHashReducer hash_reduce) const {
+    hash_reduce(inputs);
+    hash_reduce(sampling_params);
+  }
+
+  static constexpr const char* _type_key = "mlc.serve.Request";
+  static constexpr const bool _type_has_method_sequal_reduce = true;
+  static constexpr const bool _type_has_method_shash_reduce = true;
+  TVM_DECLARE_BASE_OBJECT_INFO(RequestNode, Object);
+};
+
+class Request : public ObjectRef {
+ public:
+  explicit Request(Array<Input> inputs, String sampling_params_json);
+
+  TVM_DEFINE_OBJECT_REF_METHODS(Request, ObjectRef, RequestNode);
+};
+
+/****************** Request State ******************/
+
+/*!
+ * \brief The state of a request with regard to some single model.
+ * \details In MLC LLM, the serving engine may leverage multiple models
+ * to fulfill a user generation request (e.g., use speculation decoding).
+ * For each request, we isolate its states (e.g. the generated tokens)
+ * on each model. This is to say, we use RequestModelState to store
+ * the state of a user request on a single model (rather than all models).
+ */
+class RequestModelStateNode : public Object {
```

Comment:
This is an internal state to the engine so it would be great to put it in a separate file, e.g. request_state.h or simply engine.h

Reply:
For customized logit-compute functions, they will need to take the state as a parameter to get the committed tokens I think? In this case we need to pass this class's instance. Perhaps we can have a better name for it.

Or the other possibility is we define another data class for passing the token history to the logic-compute function. And in this way we can keep RequestModelState as an internal one.

Reply:
I can move both `class RequestModelStateNode` and `struct RequestState` to a new file indeed BTW.

Reply:
Main thing is to separate it from request(which is more user facing). your proposed action works